### PR TITLE
SCB-1081 current issue when Calling the compensation from CompositeOmegaCallback (#353)

### PR DESCRIPTION
**Pull request by takumiCX**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by takumiCX** [comment]
#353

**Comment by zhfeng** [comment]
@takumiCX thanks for your contribute on the servicecomb !

Can you add an unit test to reproduce this issue ? I think this could be very useful for us to understand it. Also would you please to raise an JIRA for this issue on https://issues.apache.org/jira/browse/SCB

File: alpha/alpha-core/src/main/java/org/apache/servicecomb/saga/alpha/core/CompositeOmegaCallback.java [hunk 1]
```diff
@@ -37,14 +39,15 @@ public CompositeOmegaCallback(Map<String, Map<String, OmegaCallback>> callbacks)
...
-    if (serviceCallbacks.isEmpty()) {
-      throw new AlphaException("No such omega callback found for service " + event.serviceName());
-    }
-
     OmegaCallback omegaCallback = serviceCallbacks.get(event.instanceId());
     if (omegaCallback == null) {
       LOG.info("Cannot find the service with the instanceId {}, call the other instance.", event.instanceId());
-      omegaCallback = serviceCallbacks.values().iterator().next();
+
+      Iterator<OmegaCallback> iterator = new ArrayList<>(serviceCallbacks.values()).iterator();
```
[the remark this answers is not in the document]
Comment (takumiCX) [reply alpha/alpha-core/src/main/java/org/apache/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: 这种并发异常本来出现的概率就非常小,目前像我这么改代价基本上是最小的,我还想过通过同步的方式,不过这样可能会稍微影响性能,而且也不利于后期维护。你说的用户选择服务实例是不是有点把简单问题搞复杂了呢?

**Comment by takumiCX** [comment]
> @takumiCX thanks for your contribute on the servicecomb !
> 
> Can you add an unit test to reproduce this issue ? I think this could be very useful for us to understand it. Also would you please to raise an JIRA for this issue on https://issues.apache.org/jira/browse/SCB



> @takumiCX thanks for your contribute on the servicecomb !
> 
> Can you add an unit test to reproduce this issue ? I think this could be very useful for us to understand it. Also would you please to raise an JIRA for this issue on https://issues.apache.org/jira/browse/SCB



> @takumiCX thanks for your contribute on the servicecomb !
> 
> Can you add an unit test to reproduce this issue ? I think this could be very useful for us to understand it. Also would you please to raise an JIRA for this issue on https://issues.apache.org/jira/browse/SCB

```
public void compensate(TxEvent event) {
 Map<String, OmegaCallback> serviceCallbacks = callbacks.getOrDefault(event.serviceName(), emptyMap());
if (serviceCallbacks.isEmpty()) {
 throw new AlphaException("No such omega callback found for service " + event.serviceName());
 }
OmegaCallback omegaCallback = serviceCallbacks.get(event.instanceId());
 if (omegaCallback == null) {
 LOG.info("Cannot find the service with the instanceId {}, call the other instance.", event.instanceId());
 omegaCallback = serviceCallbacks.values().iterator().next();
 }
try {
 omegaCallback.compensate(event);
 } catch (Exception e) {
 serviceCallbacks.values().remove(omegaCallback);
 throw e;
 }
 }
```

线程可能通过了 if (omegaCallback == null) 的判断条件但是在omegaCallback = serviceCallbacks.values().iterator().next()之前失去了cpu执行权,由于其他线程对serviceCallbacks这个map的操作,目前看alpha端有两种情况:一种是接收到omega端的onDisconnected请求将对应omega端实例从map中移除;一种是执行pendingTask的线程重新进行补偿时失败执行下面这部分代码catch (Exception e)

{ serviceCallbacks.values().remove(omegaCallback); throw e; }

时也会移除map中对应的omega端实例。

我觉得我在issue中分析的已经够清晰了。这种并发异常发生的可能性非常小,我是发现了可能存在的bug后通过编写例子验证了我的想法,不过需要在源码上做些改动,这种比较容易复现。

```
if (omegaCallback == null) {
 LOG.info("Cannot find the service with the instanceId {}, call the other instance.", event.instanceId());
 try {
 TimeUnit.SECONDS.sleep(2);
 } catch (InterruptedException e) {
 e.printStackTrace();
 }

 omegaCallback = serviceCallbacks.values().iterator().next();

}
 
try {

 throw new RuntimeException();
// omegaCallback.compensate(event);
 } catch (Exception e) {
```

然后下面是我写的测试代码
```
@Test
  public void compensateWithConcurrency() throws InterruptedException {

    ConcurrentHashMap<String, OmegaCallback> serviceCallbacks = new ConcurrentHashMap();
    serviceCallbacks.put(instanceId1One,callback1One);
    callbacks.put(serviceName1,serviceCallbacks);
    new Thread(new Runnable() {
      @Override
      public void run() {
        compositeOmegaCallback.compensate(eventOf(serviceName1,instanceId1Two,TxStartedEvent));
      }
    }).start();


    TimeUnit.SECONDS.sleep(1);

    new Thread(new Runnable() {
      @Override
      public void run() {
        compositeOmegaCallback.compensate(eventOf(serviceName1,instanceId1One,TxStartedEvent));
      }
    }).start();

    TimeUnit.SECONDS.sleep(3);

  }

```

File: alpha/alpha-core/src/main/java/org/apache/servicecomb/saga/alpha/core/CompositeOmegaCallback.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (WillemJiang) [reply alpha/alpha-core/src/main/java/org/apache/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: We could provide a default implementation and user change it if they want.

**Pull request merged**